modules/data-streams/src/internalClusterTest/java/org/elasticsearch/datastreams/logsdb/qa/AbstractChallengeTest.java

```diff
@@ -0,0 +1,281 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.datastreams.logsdb.qa;
+
+import org.elasticsearch.action.DocWriteRequest;
+import org.elasticsearch.action.admin.indices.refresh.RefreshRequest;
+import org.elasticsearch.action.admin.indices.template.put.TransportPutComposableIndexTemplateAction;
+import org.elasticsearch.action.bulk.BulkRequest;
+import org.elasticsearch.action.bulk.BulkResponse;
+import org.elasticsearch.action.datastreams.CreateDataStreamAction;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.action.search.SearchResponse;
+import org.elasticsearch.action.support.broadcast.BroadcastResponse;
+import org.elasticsearch.action.support.master.AcknowledgedResponse;
+import org.elasticsearch.cluster.metadata.ComposableIndexTemplate;
+import org.elasticsearch.cluster.metadata.Template;
+import org.elasticsearch.common.CheckedSupplier;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.compress.CompressedXContent;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.core.CheckedConsumer;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.datastreams.DataStreamsPlugin;
+import org.elasticsearch.plugins.Plugin;
+import org.elasticsearch.search.builder.SearchSourceBuilder;
+import org.elasticsearch.test.ESSingleNodeTestCase;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentFactory;
+import org.junit.Before;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * This class encapsulates the logic for running tests using two data streams to compare results
+ * from indexing and querying operations. The `oracle` data stream serves as the source of truth,
+ * while the `challenge` data stream is tested against the oracle. Comparing operation results
+ * on both data streams ensures that the challenge data stream behaves as expected relative to the
+ * oracle data stream.
+ */
+public abstract class AbstractChallengeTest extends ESSingleNodeTestCase {
+    private final String oracleDataStreamName;
+    private final String challengeDataStreamName;
+    private final String oracleTemplateName;
+    private final String challengeTemplateName;
+
+    // NOTE: this is a single node test, having more shards or replicas results in a yellow cluster
+    private static final int DEFAULT_NUMBER_OF_SHARDS = 1;
+    private static final int DEFAULT_NUMBER_OF_REPLICAS = 1;
```

Comment:
I will probably change the naming here and in other places to use `expected` and `actual`.